CircleCI 2.0 fast-finish fix

jakirkham opened the pull request:
Fixes https://github.com/conda-forge/conda-forge-build-setup-feedstock/issues/75

Makes a tweak to the CircleCI fast-finish code so that it will correctly filter builds based on matrix item (if there is one). Previously a build could be ended prematurely if it wasn't the latest build even if it was for a different matrix case (e.g. Python version). This filters builds matching the pull request by the matrix case first. That way if there is a newer build number for a different matrix case, it doesn't factor into the determination of whether the current build should run or not.

cc @jjhelmus

jakirkham committed:
Filter CircleCI builds for job name matches

For workflow builds the build number can increase, but the job name may
be different (especially if there is a matrix). So before comparing
build numbers to determine if a build is out-of-date, filter the builds
to find all the jobs that share the same job name.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

jakirkham commented on recipe/ff_ci_pr_build.py:
```diff
@@ -61,7 +61,12 @@ def circle_check_latest_pr_build(repo, pr, build_num):
-    pr_build_nums = set(map(lambda b: int(b["build_num"]), builds))
+    job_name = os.environ.get("CIRCLE_JOB")
+    same_param_builds = list(filter(
+        lambda b: b.get("build_parameters", {}).get("CIRCLE_JOB") == job_name,
```
For some reason, the default value of this `get` method for `"build_parameters"` is not being used. 😕 Not entirely sure why that is happening, but it does happen to me locally as well. Suggestions welcome.

jakirkham replied:
Have added a workaround for this issue whatever it may have been.

jakirkham committed:
Write out parameter check completely

Seems there is some bug with Python `dict`'s `get` method, which we are
running into here. In particular, it seems to ignore the `default` value
provided to it. AFAICT it seems to be related to the key we are
providing (maybe length related :/). This additional code acts as a
workaround for this bug to ensure we are able to handle the pre-CircleCI
2.0 case.

jakirkham committed:
Bump to 4.4.9

Need a new release to demarcate the change to the fast finish script
w.r.t. CircleCI.

jakirkham commented:
As can be seen by this [outdated build]( https://circleci.com/gh/conda-forge/conda-forge-build-setup-feedstock/198 ) and this [newer build]( https://circleci.com/gh/conda-forge/conda-forge-build-setup-feedstock/199 ), the pre-CircleCI 2.0 case is still handled correctly.

jakirkham committed:
MNT: Re-rendered with conda-smithy 2.3.3
